### PR TITLE
Update Rule “do-you-have-separate-development-testing-and-production-environments/rule”

### PR DESCRIPTION
**Pull request by danielmackay**
> 1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)

✏️Email - Rule Update - Do you have separate development, testing and production environments?

This was an outdated rule, that didn't actually proposed what environments were recommended and what their different usages were for.

The idea is to propose a standard set of environments and define what each one means.

> 2. What was changed?

✏️Identified names for different environments and explained the differences between them.  Also recommended, combinations of environments that can be used for smaller vs larger projects.

The important part (but not the entire rule):
image

> 3. Did you do pair or mob programming?

✏️ No

### Notes
- https://learn.microsoft.com/en-us/aspnet/core/fundamentals/environments?view=aspnetcore-8.0

### CODE REVIEW

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md [hunk 1]
```diff
@@ -1,28 +1,58 @@
...
+
+* **Production**: Real data being used by real customers
+* **Staging**: 'Production-like' data.  Used for testing and verification before deploying to Production.  Should have 'production-like' access (i.e. avoid giving devs admin rights).  Should generally have the same specs as production (especially during performance testing).  However, this is not always the case due to cost implications.
+* **Test**: Internal environment used by the development team.  Hosted in the same place as the Staging & Production environments.  Access usually not provided to the client.
+* **Ephemeral**: Special environment that is spun up on demand for testing of a branch.
+* **Development**: Developer environment running on their local machine.  Also, sometimes called 'Local'.
+
+![Figure: Environment flow](ssw-rules-environments.jpg "Environment flow")
+
+# Large or Multi-Team Projects
```
Comment (bradystroud) [inline rules/do-you-have-separate-development-testing-and-production-environments/rule.md:35]:
```suggestion
## Large or Multi-Team Projects
```
Can you make these H2s? There should only be one H1 per page (for SEO)

**Review by bradystroud — CHANGES_REQUESTED** [review]

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md [hunk 2]
```diff
@@ -1,80 +1,83 @@
...
 <!--endintro-->
 
-::: greybox
-The old solution is to use build scripts (.bat and .vbs files) to automatically create a setup package that can be used to deploy to testing and production environments. For backend changes, you can either include the change scripts with the setup package (if it's a localised database), or run those scripts as part of your deployment process.
+## What is each environment for?
 
-Read more about setup packages at [SSW's Wise Standard for Products.](http://www.ssw.com.au/ssw/Standards/wisesetup/WiseStandards.aspx)
-:::
+* **Production**: Real data being used by real customers
+* **Staging**: 'Production-like' data.  Used for testing and verification before deploying to Production.  Should have 'production-like' access (i.e. avoid giving devs admin rights).  Should generally have the same specs as production (especially during performance testing).  However, this is not always the case due to cost implications
```
Comment (wicksipedia) [inline rules/do-you-have-separate-development-testing-and-production-environments/rule.md:28]:
```suggestion
* **Staging**: 'Production-like' environment used for final sign-off.  Used for testing and verification before deploying to Production.  Should be as close to Production as possible e.g. access (avoid giving devs admin rights), same specs as production (especially during performance testing).  However, this is not always the case due to cost implications
```

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md [hunk 3]
```diff
@@ -1,80 +1,83 @@
...
+* **Test**: Internal environment used by the development team.  Hosted in the same place as the Staging & Production environments.  Access usually not provided to the client
```
Comment (wicksipedia) [inline rules/do-you-have-separate-development-testing-and-production-environments/rule.md:29]:
```suggestion
* **Test**: A place to verify code changes. Typically simpler or under-specified version of Staging or Production environments aiding in the early identification and troubleshooting of issues.
```
> Internal environment used by the development team.
> Access usually not provided to the client
I'm not sure about these comments, why does it need to be so prescriptive

Comment (william-liebenberg) [reply rules/do-you-have-separate-development-testing-and-production-environments/rule.md:29]: The client still has access to the dev/test environments (they're paying for it). We give them the understanding that this environment is for the dev team's test purposes and that there is no guarantee that this is a stable environment.

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md [hunk 4]
```diff
@@ -1,80 +1,83 @@
...
+* **Ephemeral**: Special environment that is spun up on demand for testing of a branch
```
Comment (william-liebenberg) [inline rules/do-you-have-separate-development-testing-and-production-environments/rule.md:30]: Short-lived environment that is spun up on demand for isolated testing of a branch, and then torn down when the branch is merged or deleted.

We have a rule for ephemeral environments that you can link to: https://www.ssw.com.au/rules/ephemeral-environments/

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md [hunk 5]
```diff
@@ -1,80 +1,83 @@
...
-The best way to manage your product across different environments is to use DevOps tools (like Azure DevOps, GitHub Actions, and others) and Infrastructure as Code (IaC).
-    
-For more information, see our rule: [Do you know how to create Azure resources?](/azure-resources-creating).
-:::
+### Large or Multi-Team Projects
 
-::: good
-Good Example - use DevOps tools and IaC to deploy to different environments
-:::
+For large projects it's recommended to run 3 hosted environments + 1 local:
```
Comment (william-liebenberg) [inline rules/do-you-have-separate-development-testing-and-production-environments/rule.md:37]: Most of the projects over the years have stuck to the following designations:

Hosted: 
* Production (prod)
* Staging (stage)/(staging)
* Devlopment (dev)

Local:
* Local (localhost)

I'd recommend sticking to the list above to be consistent with our real-life projects.

There has been some projects with QA, UA, Test, SIT, PP, PreProd environment designations. Perhaps your rule can include a statement that starting with the simple set of environments is good, and then there can be more environments added when needed for specific processes.

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md (hunk 3 above)
Comment (danielmackay) [reply rules/do-you-have-separate-development-testing-and-production-environments/rule.md]: Have removed the comment about client access to 'Test' as per the suggestion from @wicksipedia.

File: rules/do-you-have-separate-development-testing-and-production-environments/rule.md (hunk 4 above)
Comment (danielmackay) [reply rules/do-you-have-separate-development-testing-and-production-environments/rule.md]: Have updated as per your comment.

**Comment by danielmackay** [comment]
As per my conversation with @wicksipedia, we have decided to leave this rule as is and create a new rule titled "do you know which environments you need to provision when starting a new project?"

**Pull request closed without merging**